src/executor/thread_factory.cpp

```diff
@@ -23,40 +40,310 @@ class Thread final : public IThread {
+
+ protected:
+  ThreadFactory& GetThreadFactory() override {
+    return *this;
+  }
+  size_t GetPriority() const override {
+    return 0;
+  }
+  std::string_view GetName() const override {
+    return "";
+  }
+  Functor GetAcquire() const override {
+    return nullptr;
+  }
+  Functor GetRelease() const override {
+    return nullptr;
+  }
+
+  const size_t _max_threads;
+};
+
+class LightThreadFactory final : public ThreadFactory {
+ public:
+  explicit LightThreadFactory(size_t max_threads) : ThreadFactory{max_threads} {
+  }
+
+ private:
+  IThreadPtr Acquire(Functor functor, size_t priority, std::string_view name,
+                     Functor acquire, Functor release) final {
+    if (_threads.fetch_add(1, std::memory_order_acq_rel) < _max_threads) {
+      return std::make_unique<LightThread>(priority, name, std::move(functor),
+                                           std::move(acquire),
+                                           std::move(release));
+    }
+    // It's fast and ensures that we never create more than _max_threads
+    // std::thread, but it doesn't guarantee that we won't get nullptr back
+    // when _threads < _max_threads.
+    _threads.fetch_sub(1, std::memory_order_relaxed);
     return nullptr;
   }
 
   void Release(IThreadPtr thread) final {
-    _threads_count.fetch_sub(1, std::memory_order_relaxed);
+    static_cast<LightThread&>(*thread).Join();
     thread.reset();
+    _threads.fetch_sub(1, std::memory_order_release);
   }
 
-  std::atomic<size_t> _threads_count{0};
-  const size_t _cached_threads_count;
-  const size_t _max_threads_count;
+  std::atomic_size_t _threads{0};
+};
+
+class CachedThreadFactory final : public ThreadFactory {
+ public:
+  CachedThreadFactory(size_t cache_threads, size_t max_threads)
+      : ThreadFactory{max_threads}, _cache_threads{cache_threads} {
+  }
+
+ private:
+  IThreadPtr Acquire(Functor functor, size_t priority, std::string_view name,
+                     Functor acquire, Functor release) final {
+    std::unique_lock guard{_m};
+    if (_threads_count == _max_threads) {
+      return nullptr;
+    }
+    IThreadPtr thread{_threads.PopBack()};
+    if (thread == nullptr) {
+      thread = std::make_unique<CachedThread>();
+      ++_threads_count;
+    }
+    guard.unlock();
+
+    static_cast<CachedThread&>(*thread).Set(priority, name, std::move(functor),
+                                            std::move(acquire),
+                                            std::move(release));
+    return thread;
+  }
+
+  void Release(IThreadPtr thread) final {
+    std::unique_lock guard{_m};
+    if (_threads_count < _cache_threads) {
+      _threads.PushBack(thread.release());
+    } else {
+      static_cast<CachedThread&>(*thread).Join();
+      thread.reset();
+      --_threads_count;
+    }
+  }
+
+  const size_t _cache_threads;
+
+  std::mutex _m;
+  size_t _threads_count{0};
+  container::intrusive::List<IThread> _threads;
```

Comment:
You can try using an MPMC lock-free queue of fixed size. 
But the main problem is that then threads will wake up and fall asleep in FIFO and not LIFO order